### PR TITLE
Recurring Payments Block: Minimum transaction amounts

### PR DESCRIPTION
**Pull request by beaucollins**
Uses Stripe's minimum transaction amounts per currency.

#### Changes proposed in this Pull Request:

- Uses [Stripe's minimum transaction amounts per currency](https://stripe.com/docs/currencies#minimum-and-maximum-charge-amounts).
- Sets the input minimum value to 0 to allow for better stepper experience
- Prevents using less than minimum amount during component state change
- Shows the minimum price for the selected currency below the Price input field.

The placeholder text uses `formatCurrency` which is nice, however it somewhat misleads the user into thinking they can type the currency symbols into the field while we actually only allow that which can be parsed by `parseFloat` which means values like `$0.50` or `CHF 0,50` are not considered valid. Should we allow this kind of input?

#### Testing instructions:

This assumes the "Recurring Payments Block" is available to your Jetpack plan for the site you are testing.

##### Setup

* Go to "Posts → Add New"
* Add a block and type `/recurr` select "Recurring Payments button"
   image
* If a plan already exists press the "Add a plan" button to display the form:
  image

##### Test

* Clear out the "price" value, the placeholder should show the minimum price
  image

* Change the currency, if the price is now too low for the given currency, it should show the error (e.g. start with `USD` and `0.5` and change to `MXN`)
  image

#### Proposed changelog entry for your changes:
* Recurring Payments block minimum price is updated and is dependent on selected currency.

### CODE REVIEW

File: extensions/blocks/recurring-payments/index.js
Comment (beaucollins) [inline extensions/blocks/recurring-payments/index.js]: @artpi Should we default to a value if the currency is not one of the known options? This shouldn't happen in practice.

File: extensions/blocks/recurring-payments/edit.jsx
Comment (jeherve) [inline extensions/blocks/recurring-payments/edit.jsx]: Do you think you could rely on `sprintf` here and include the price in the translatable string? I think that would make things easier for translators.

**Review by jeherve — DISMISSED** [review]

File: extensions/blocks/recurring-payments/edit.jsx
Comment (beaucollins) [reply extensions/blocks/recurring-payments/edit.jsx]: Yes, was looking for the recommendation. [The handbook was a bit ambiguous as to handle interpolation](https://developer.wordpress.org/block-editor/developers/internationalization/).

File: extensions/blocks/recurring-payments/index.js
Comment (artpi) [inline extensions/blocks/recurring-payments/index.js]: Can we combine this with `SUPPORTED_CURRENCY_LIST` ?
Something like 
```
export const SUPPORTED_CURRENCY_LIST = [
	'USD': { 'minimumCharge': 0.5 },
```

I am afraid of a situation where somebody adds a new currency without introducing minimum charge

File: extensions/blocks/recurring-payments/index.js
Comment (artpi) [reply extensions/blocks/recurring-payments/index.js]: The currency list gets populated from these constants above. So that should not be happening and we don't want want fallbacks - we want errors in this case :)

**Review by artpi — DISMISSED** [review]
My comments are not huge blockers

File: extensions/blocks/recurring-payments/index.js
Comment (beaucollins) [reply extensions/blocks/recurring-payments/index.js]: There is now a single list. One thing to note, this change now adds more supported currencies: `INR`, `MYR`, and `PLN`. I don't know if that's desired.

**Review by artpi — CHANGES_REQUESTED** [review]
In saveProduct function there is still `this.state.editedProductPrice < 5` which does not allow to actually save these products:
Zrzut ekranu 2020-02-28 o 10 28 17

**Comment by beaucollins** [comment]
> In saveProduct function there is still this.state.editedProductPrice < 5 which does not allow to actually save these products:

Good catch @artpi, updated in fc53ad2.

**Comment by beaucollins** [comment]
@jeherve @artpi changes have been addressed, any chance I can get another look?

**Review by artpi — DISMISSED** [review]
Looks good! I am most excited about PLN subscription! 👍

**Comment by beaucollins** [comment]
@jeherve you made a blocking change request last week, any chance you can take another look?

If you aren’t going to follow up perhaps make your change requests non-blocking?

**Comment by kraftbj** [comment]
@beaucollins Someone Crew (jeherve or another Crewmate) needs to approve before merge either way.

**Comment by kraftbj** [comment]
Before approving, as it works as advertised for a new recurring payment plan, I'm curious about the migration plan to existing invalid recurring payment blocks.

For example, if on 8.3 before this PR, I add a plan for MX$5 (where the minimum is MX$10), after upgrade, there is seemingly no notification that I have an invalid plan active on my site or ability to update it.

I can add a new block and select anew the existing, invalid plan without issue.
2020-03-04 at 6 37 PM

**Comment by beaucollins** [comment]
@artpi can you chime in about the question:

> For example, if on 8.3 before this PR, I add a plan for MX$5 (where the minimum is MX$10), after upgrade, there is seemingly no notification that I have an invalid plan active on my site or ability to update it.

**Comment by artpi** [comment]
> I'm curious about the migration plan for existing invalid recurring payment blocks.
> For example, if on 8.3 before this PR, I add a plan for MX$5 (where the minimum is MX$10), after upgrade, there is seemingly no notification that I have an invalid plan active on my site or ability to update it.


There is no migration path because they are plainly not working. If you add a MX$5 , the transaction will fail during checkout, with an error that is confusing to the user.
This whole PR is intended to fix this problem.

There are 2 users with plans in MXN<10  and 4 users with plans in JPY that have this issue.

To make sure everything is proper in the block, let's filter out the invalid subscriptions (make them impossible to select)

We will not deal with those 6 users. Their subscriptions are not working anyway, by definition these plans have no subscribers, so they can retry and create a new plan.

**Comment by kraftbj** [comment]
Making invalid plans impossible to select and not providing a fix for those six is reasonable to me.

If you could Slack me those users or how to pull that list, I'm happy to reach out to them proactively to let them know they need to amend their subscriptions. That'd be a nice priority support outreach.

**Comment by beaucollins** [comment]
@kraftbj @artpi  membership options that do not meet the minimum threshold are no longer available to be selected.

Currently they are filtered out as opposed to "shown and disabled".

**Review by kraftbj — APPROVED** [review]
Thanks for the updates! Let's do this.

**Comment by artpi** [comment]
Hey @beaucollins I merged something and suspect you will have to:
- rebase this
- wait for D39465-code to pick up new changes (otherwise it will have conflicts )
- get new approval from this from Brandon
- merge D39465-code
- Merge this

**Comment by beaucollins** [comment]
D39465 is deployed.

**Pull request merged**
